Functions ASB sample update (#5110)
Particular/docs.particular.net: samples/previews/azure-functions/service-bus/sample.md

---
title: Using NServiceBus in Azure Functions with Service Bus triggers
reviewed: 2020-11-10
component: ASBFunctions
related:
 - previews/azure-functions-service-bus
---

This sample shows how to host NServiceBus within an Azure Function, in this case, a function triggered by an incoming Service Bus message. This enables hosting message handlers in Azure Functions, gaining the abstraction of message handlers implemented using `IHandleMessages<T>` and also taking advantage of NServiceBus's extensible message processing pipeline.

> The sample demonstrates two approaches that achieve the same outcome:

Reviewer: ```suggestion
The sample demonstrates two configuration approaches that achieve the same outcome:
```

> 1. Using `IFunctionHostBuilder` and Microsoft DI container approach

Reviewer: ```suggestion
1. Integrating with the `IFunctionHostBuilder`, including the host managed DI container.
```

> 2. Static approach

Reviewer: ```suggestion
2. Configuring the endpoint inside the trigger class as static field.
```

az servicebus queue create --name ASBTriggerQueue --namespace-name <asb-namespace-to-use> --resource-group <resource-group-containing-namespace>
```

To use the sample, a valid Service Bus connection string must be provided in the `local.settings.json` file.

## Sample structure

The sample contains the following projects:
- `AzureFunctions.Sender` - console application to generate a trigger message

> - `AzureFunctions.ASBTrigger.FunctionsHostBuilder` - implemenetation using `IFunctionHostBuilder` approach to host NServiceBus endpoint

Reviewer: ```suggestion
- `AzureFunctions.ASBTrigger.FunctionsHostBuilder` - Using `IFunctionHostBuilder` approach to host the NServiceBus endpoint
```

> - `AzureFunctions.ASBTrigger.Static` - implementation using static approach to host NServiceBus endpoint

Reviewer: ```suggestion
- `AzureFunctions.ASBTrigger.Static` - Using static approach to host the NServiceBus endpoint
```

## Code walk-through

### `IFunctionHostBuilder` approach

The NServiceBus endpoint configured using `IFunctionHostBuilder` is using the convention and is wired using `Startup` class like this:

snippet: configuration-with-function-host-builder

> `FunctionEndpoin` is then injected into the function class:

Reviewer: ```suggestion
`FunctionEndpoint` is then injected into the function class:
```

Meanwhile, the message handlers for `TriggerMessage` and `FollowUpMessage`, also hosted within the Azure Functions project, are normal NServiceBus message handlers, which are also capable of sending messages themselves.

### Handlers

> Both approaches use the same handlers code, with a `CustomDependency` passed in.

Reviewer: ```suggestion
Both approaches use the same message handlers, with a `CustomDependency` passed in.
```
